src/permutations_const.rs

```diff
@@ -0,0 +1,188 @@
+use alloc::boxed::Box;
+use alloc::vec::Vec;
+use std::fmt;
+use std::iter::once;
+use std::iter::FusedIterator;
+use std::convert::TryInto;
+
+use super::lazy_buffer::LazyBuffer;
+use crate::size_hint::{self, SizeHint};
+
+/// An iterator adaptor that iterates through all the `k`-permutations of the
+/// elements from an iterator.
+///
+/// See [`.permutations()`](crate::Itertools::permutations) for
+/// more information.
+#[must_use = "iterator adaptors are lazy and do nothing unless consumed"]
+pub struct PermutationsConst<I: Iterator, const K: usize> {
+    vals: LazyBuffer<I>,
+    state: PermutationState<K>,
+}
+
+impl<I, const K: usize> Clone for PermutationsConst<I, K>
+where
+    I: Clone + Iterator,
+    I::Item: Clone,
+{
+    clone_fields!(vals, state);
+}
+
+#[derive(Clone, Debug)]
+enum PermutationState<const K: usize> {
+    /// No permutation generated yet.
+    Start,
+    /// Values from the iterator are not fully loaded yet so `n` is still unknown.
+    Buffered { min_n: usize },
+    /// All values from the iterator are known so `n` is known.
+    Loaded {
+        indices: Box<[usize]>,
+        cycles: Box<[usize]>,
+    },
+    /// No permutation left to generate.
+    End,
+}
+
+impl<I, const K: usize> fmt::Debug for PermutationsConst<I, K>
+where
+    I: Iterator + fmt::Debug,
+    I::Item: fmt::Debug,
+{
+    debug_fmt_fields!(Permutations, vals, state);
+}
+
+pub fn permutations_const<I: Iterator, const K: usize>(iter: I) -> PermutationsConst<I, K> {
+    PermutationsConst {
+        vals: LazyBuffer::new(iter),
+        state: PermutationState::Start,
+    }
+}
+
+impl<I, const K: usize> Iterator for PermutationsConst<I, K>
+where
+    I: Iterator,
+    I::Item: Clone + Default,
+{
+    type Item = [I::Item; K];
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let Self { vals, state } = self;
+        match state {
+            PermutationState::Start => {
+                *state = PermutationState::End;
+                // TODO: Consider this case and handle it somehow, currently just using default
+                Some(std::array::from_fn(|_|I::Item::default()))
+            }
```

Comment:
Not sure how to proceed about this, from what i can tell the API guarantees the vector to always be K elements long, but the reference returns an empty vec?